feat(app) an AdminPage for admin plugins

gashcrumb opened the pull request:
## Description

This change adds a Administration navigation tab and page intended for plugins that are for administration of an Janus/RHDH instance.  This commit adds two routes /admin/rbac and /admin/plugins and related mountpoints so that dynamic plugins can contribute UI components to this page.  The page can also deal with a few edge cases and will also not be visible should no plugins be configured to show up on it.  A test is provided that covers these cases.  This change also adds the dynamic-plugins-info dynamic plugin which is one of two components intended to appear on this page.

## Which issue(s) does this PR fix

Part of [RHIDP-980](https://issues.redhat.com/browse/RHIDP-980).  This is an approach to provide a home for dynamic plugins that fall under the "Administration" tab, for example:

![image](https://github.com/janus-idp/backstage-showcase/assets/351660/9ae23e90-5e9d-462c-a684-517792a1684c)

## PR acceptance criteria

Please make sure that the following steps are complete:

- [ ] GitHub Actions are completed and successful
- [ ] Unit Tests are updated and passing
- [ ] E2E Tests are updated and passing
- [ ] Documentation is updated if necessary (requirement for new features)
- [x] Add a screenshot if the change is UX/UI related

## How to test changes / Special notes to the reviewer
The dynamic-plugins-info plugin is added to the showcase

To enable the Administration tab and see the installed dynamic plugins, use the following configuration in your `app-config.local.yaml` to enable the "Administration" nav and "Plugins" tab:

```yaml
dynamicPlugins:
  frontend:
    janus-idp.backstage-plugin-dynamic-plugins-info:
      dynamicRoutes:
        - path: /admin/plugins
          importName: DynamicPluginsInfo
      mountPoints:
        - mountPoint: admin.page.plugins/cards
          importName: DynamicPluginsInfo
          config:
            layout:
              gridColumn: "1 / -1"
              width: 100vw
```

It should be possible to browse the list of installed plugins using the pagination arrows:

![image](https://github.com/janus-idp/backstage-showcase/assets/351660/f292d41d-fe51-4587-8bba-92ff6d2d3af2)

and searching should work:

![image](https://github.com/janus-idp/backstage-showcase/assets/351660/281e2c01-919b-4985-9720-c64e20161b08)

Removing the configuration for the tab/plugin shouldn't leave the user with a blank page:

![image](https://github.com/janus-idp/backstage-showcase/assets/351660/5dccd59c-a7c1-4e02-a293-1dcea5cf73f2)

invincibleJai commented on packages/app/src/components/AppBase/AppBase.tsx:
```diff
@@ -64,6 +65,16 @@ const AppBase = () => {
             <Route path="/catalog-graph" element={<CatalogGraphPage />} />
             <Route path="/learning-paths" element={<LearningPaths />} />
+            <Route path="/admin" element={<AdminPage />} />
+            {dynamicRoutes
```
as you are handling the routes here for admin and below there is a map run on all dynamicRoutes again so can that be an issue?

https://github.com/janus-idp/backstage-showcase/pull/932/files#diff-4ecb4d92de7b045a6d54c1e7327fb3f2e41f34316fd87dfa01a629d08a93f971R78-R88

gashcrumb replied:
This I have to revisit, as I figured I should be able to [just map these pages to the admin route](https://github.com/janus-idp/backstage-showcase/pull/932/files#diff-4ecb4d92de7b045a6d54c1e7327fb3f2e41f34316fd87dfa01a629d08a93f971R70) and then filter them from the other block (it would be before [this ](https://github.com/janus-idp/backstage-showcase/pull/932/files#diff-4ecb4d92de7b045a6d54c1e7327fb3f2e41f34316fd87dfa01a629d08a93f971L68)`map`)  where these are mounted.  But I was surprised to find it didn't work at the time, so yep this needs a closer look.  I might have just confused myself with my test plugin :smile:

invincibleJai commented on packages/app/src/components/admin/AdminPage.tsx:
```diff
@@ -0,0 +1,87 @@
+    );
+  }
+  if (invalidTab(selectedTab)) {
+    // in this case the page will navigate to a tab, so avoid
```
which tab will it navigate to? in this case? and if am getting it right currently on rbac and plugins tab would be allowed right? in future, if any other tab needs to be added then it will require code change in this file?

gashcrumb replied:
Correct.  So this page will either navigate to the "rbac" tab or the "plugin" tab depending on what's available, and if there's no configured tab, for example let's say the configuration for all of this is removed while a user is at one of these pages then the page resorts to showing the standard error page which provides the user a link out of this area.

gashcrumb replied:
Oh yes, and for this page additional tabs can be added by modifying [this array](https://github.com/janus-idp/backstage-showcase/pull/932/files#diff-1913f62d77a8135d4b4a3c8abd92927824cf8fbd41f5e94b098880092229257cR14-R25)

invincibleJai commented on packages/app/src/components/admin/AdminPage.tsx:
```diff
@@ -0,0 +1,87 @@
+  },
+];
+
+export const AdminPage = () => {
```
good to add some unit tests for this. Can be taken up as follow up PR if needed

gashcrumb replied:
yep for sure

gashcrumb replied:
https://github.com/janus-idp/backstage-showcase/issues/940 unless I finish it before this PR is merged.

gashcrumb replied:
Added https://github.com/janus-idp/backstage-showcase/pull/932/files#diff-43884aaa2324a9c65904cce4a0137a8754deb314ccdd58112fb9a31a13bb56a9

gashcrumb commented:
Rebased, and I have updated this to handle the case where the user might be on the admin page when there's no configuration for it, and some comments in there too.

debsmita1 commented:
Verified it locally

<img width="1717" alt="Screenshot 2024-02-07 at 12 47 04 AM" src="https://github.com/janus-idp/backstage-showcase/assets/22490998/340046d2-2f8d-44bc-af98-3c657d369cb9">

@gashcrumb Would you mind taking a look at the code smells ?

gashcrumb commented:
> @gashcrumb Would you mind taking a look at the code smells ?

Ah, of course, didn't spot that.  Thanks!

gashcrumb commented:
Code smells should be sorted now...

invincibleJai reviewed (approved):
/lgtm

invincibleJai commented:
Thanks @gashcrumb, this looks good :)
